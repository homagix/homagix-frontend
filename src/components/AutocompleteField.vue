<script setup lang="ts">
import { computed, ref } from "vue"

type Item = { id: string; name: string }

const props = defineProps<{
  id: string
  list: Item[]
}>()

const emit = defineEmits<{
  (event: "change", value: Item): void
}>()

const search = ref("")
const selectedIndex = ref(-1)
const suggestions = ref([] as Item[])
const listId = props.id + "-result-list"

const activeDescendant = computed(() => (selectedIndex.value >= 0 ? "" + selectedIndex.value : undefined))

function isSelected(index: number) {
  return index === selectedIndex.value
}

function selectItem(item: Item) {
  emit("change", item)
  reset()
}

const listHeight = computed(() => `${suggestions.value.length * 1.3 + 0.11}em`)

function reset() {
  suggestions.value = []
  selectedIndex.value = -1
  search.value = ""
}

function handleEnter(event: KeyboardEvent) {
  if ("Enter" === event.code) {
    if (selectedIndex.value >= 0) {
      selectListEntry()
      const item = suggestions.value.find(item => item.name === search.value)
      emit("change", item as Item)
    } else {
      emit("change", { id: "", name: search.value })
    }
    reset()
  }
}

function handleArrowKeys(event: KeyboardEvent) {
  if ("ArrowDown" === event.code) {
    selectedIndex.value = Math.min(suggestions.value.length - 1, selectedIndex.value + 1)
  } else if ("ArrowUp" === event.code) {
    selectedIndex.value = Math.max(-1, selectedIndex.value - 1)
  }
}

function selectListEntry() {
  const item = suggestions.value[selectedIndex.value]
  search.value = item.name
}

function updateSuggestions() {
  if (search.value === "") {
    reset()
  } else {
    const pattern = new RegExp(search.value, "i")
    suggestions.value = props.list.filter(item => item.name.match(pattern))
  }
}
</script>

<template>
  <div class="autocomplete" :id="id">
    <input
      type="text"
      v-model="search"
      @keydown="handleEnter"
      @keyup="handleArrowKeys"
      @input="updateSuggestions"
      role="combobox"
      autocomplete="off"
      autocapitalize="off"
      autocorrect="off"
      spellcheck="false"
      aria-autocomplete="list"
      aria-haspopup="listbox"
      :aria-owns="listId"
      :aria-activedescendant="activeDescendant"
      aria-expanded="false"
    />
    <select role="listbox" :id="listId" :size="2">
      <option
        v-for="(item, index) in suggestions"
        :key="item.id"
        :selected="isSelected(index)"
        @click="selectItem(item)"
      >
        {{ item.name }}
      </option>
    </select>
  </div>
</template>

<style lang="scss" scoped>
.autocomplete {
  display: inline-block;
  position: relative;
  box-sizing: border-box;
  width: 100%;

  input {
    width: 100%;
    font-size: 1em;
    margin-top: 2px;
  }

  input:focus + select,
  select:hover {
    display: inline;
  }

  select {
    height: v-bind(listHeight);
    display: none;
    position: absolute;
    top: 100%;
    left: 0;
    width: 100%;
    max-height: 10em;
    overflow: auto;
    background: white;
    margin: -4px 0 0;

    &:empty {
      border: none;
    }

    option {
      overflow-x: auto;
    }
  }
}
</style>
