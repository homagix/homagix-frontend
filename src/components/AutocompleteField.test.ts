import { mount } from "@vue/test-utils"
import { describe, expect, it } from "vitest"
import AutocompleteField from "./AutocompleteField.vue"

const list = [
  { id: "f", name: "first" },
  { id: "s", name: "second entry" },
  { id: "t", name: "third entry" },
]

describe("AutocompleteField", () => {
  it("should render as expected", async () => {
    const wrapper = mount(AutocompleteField, { props: { id: "field-id", list } })
    const search = wrapper.find("input")
    search.setValue("first")
    await search.trigger("input")
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should show matching entries from list after input", async () => {
    const wrapper = mount(AutocompleteField, { props: { id: "field-id", list } })
    const search = wrapper.find("input")
    await search.setValue("entry")
    await search.trigger("input")
    expect(wrapper.findAll("option").map(option => option.text())).toEqual(["second entry", "third entry"])
  })

  it("should emit a 'change' event when an entry from the list is selected and the enter key is pressed", async () => {
    const wrapper = mount(AutocompleteField, { props: { id: "field-id", list } })
    const search = wrapper.find("input")
    await search.setValue("entry")
    await search.trigger("keyup", { code: "ArrowDown" })
    await search.trigger("keydown", { code: "Enter" })
    expect(wrapper.emitted().change).toEqual([[list[1]]])
    expect(search.element.value).toBe("")
  })

  it("should emit a 'change' event when an element from the list is directly selected by clicking the mouse", async () => {
    const wrapper = mount(AutocompleteField, { props: { id: "field-id", list } })
    const search = wrapper.find("input")
    await search.setValue("entry")
    await wrapper.find("option").trigger("click")
    expect(wrapper.emitted().change).toEqual([[list[1]]])
    expect(search.element.value).toBe("")
  })

  it("should emit a 'change' event when the enter key is pressed after some text is entered", async () => {
    const wrapper = mount(AutocompleteField, { props: { id: "field-id", list } })
    const search = wrapper.find("input")
    await search.setValue("new item")
    await search.trigger("input")
    await search.trigger("keydown", { code: "Enter" })
    expect(wrapper.emitted().change).toEqual([[{ id: "", name: "new item" }]])
    expect(search.element.value).toBe("")
  })
})
