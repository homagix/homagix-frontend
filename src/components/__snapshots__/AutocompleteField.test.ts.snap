// Vitest Snapshot v1

exports[`AutocompleteField > should render as expected 1`] = `
"<div class=\\"autocomplete\\" id=\\"field-id\\" data-v-46e78e1c=\\"\\"><input type=\\"text\\" role=\\"combobox\\" autocomplete=\\"off\\" autocapitalize=\\"off\\" autocorrect=\\"off\\" spellcheck=\\"false\\" aria-autocomplete=\\"list\\" aria-haspopup=\\"listbox\\" aria-owns=\\"field-id-result-list\\" aria-expanded=\\"false\\" data-v-46e78e1c=\\"\\"><select role=\\"listbox\\" id=\\"field-id-result-list\\" size=\\"2\\" data-v-46e78e1c=\\"\\">
    <option data-v-46e78e1c=\\"\\">first</option>
  </select></div>"
`;
